README.md
# MX QR Code

![MX QR Code](/resources/img/qr_ext.png)

**MX QR Code** QR Code Generator for ExpressionEngine.

## Installation
* Place the **qr_code** folder inside your **user/addons** folder
* Go to **cp/addons** and install *MX QR Code*

## Template Tags

{exp:qr_code}

    <img src="{exp:qr_code action="" tel="" data="ExpressionEngine" title="" email="" subj=""}" alt="QR code"/>

### Parameters

*action = ""* optional

Plugin has a couple pattern for QR code messages:

* sms - send sms
* tel - phone number
* email - send email
* bm - add to bookmarks
* site - addtional data urlencode

*tel = "555-5555"* required for action type sms,tel

telephone number

*email = "name@name.com" *required for action type email

email

*subj = "John Doe"* required for action type email

subject for email

*title = "My WebSite"* required for action type bookmarks

Title for bookmark

*size = "200"* optional, default "200"

qr code size *in px*

*ecc = "M"* optional, default **M**

ECC level **L** or **M** or **Q** or **H**

*data = ""*

default data for encoding

*px_color = "000000"* optional, default **000000**

pixel color

*bk_color = "ffffff"* optional, default **ffffff**

background color

*margin = "2"* optional, default **0**

margin size in px

*logo = ""* optional

url / path to logo


*logo_size="100,100"* optional

logo size


*base_path*

path to web root (required if you use base_cache parameter)

*base_cache* optional, default "/images/cache/"

path to image folder

*base64_encode = "no"* optional

incode image to base64 to inline image


## Support Policy
This is Communite Edition (CE) add-on.

## Contributing To MX QR Code for ExpressionEngine

Your participation to MX QR Code development is very welcome!

You may participate in the following ways:

* [Report issues](https://github.com/MaxLazar/mx-qr-code/issues)
* Fix issues, develop features, write/polish documentation
Before you start, please adopt an existing issue (labelled with "ready for adoption") or start a new one to avoid duplicated efforts.
Please submit a merge request after you finish development.


## License

The MX QR Code is open-sourced software licensed under the [MIT license](http://opensource.org/licenses/MIT)


CHANGELOG.md
# MX QR Code

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## 4.0.5 - 2019-09-12
### Added
- Initial release


qr_code/qrcode/qrimage.php
<?php
/*
 * PHP QR Code encoder
 *
 * Image output of code using GD2
 *
 * PHP QR Code is distributed under LGPL 3
 * Copyright (C) 2010 Dominik Dzienia <deltalab at poczta dot fm>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

    define('QR_IMAGE', true);

    class QRimage {

        //----------------------------------------------------------------------
        public static function png($frame, $filename = false, $pixelPerPoint = 4, $outerFrame = 4,$saveandprint=FALSE, $back_color, $fore_color)
        {
            $image = self::image($frame, $pixelPerPoint, $outerFrame, $back_color, $fore_color);

            if ($filename === false) {
                Header("Content-type: image/png");
                ImagePng($image);
            } else {
                if($saveandprint===TRUE){
                    ImagePng($image, $filename);
                    header("Content-type: image/png");
                    ImagePng($image);
                }else{
                    ImagePng($image, $filename);
                }
            }

            ImageDestroy($image);
        }

        //----------------------------------------------------------------------
        public static function jpg($frame, $filename = false, $pixelPerPoint = 8, $outerFrame = 4, $q = 85)
        {
            $image = self::image($frame, $pixelPerPoint, $outerFrame, $back_color, $fore_color);

            if ($filename === false) {
                Header("Content-type: image/jpeg");
                ImageJpeg($image, null, $q);
            } else {
                ImageJpeg($image, $filename, $q);
            }

            ImageDestroy($image);
        }

        //----------------------------------------------------------------------
        private static function image($frame, $pixelPerPoint = 4, $outerFrame = 4, $back_color = 0xFFFFFF, $fore_color = 0x000000)
        {
            $h = count($frame);
            $w = strlen($frame[0]);

            $imgW = $w + 2*$outerFrame;
            $imgH = $h + 2*$outerFrame;

            $base_image = ImageCreate($imgW, $imgH);

            // convert a hexadecimal color code into decimal eps format (green = 0 1 0, blue = 0 0 1, ...)
       //     $r1 = round((($fore_color & 0xFF0000) >> 16), 5);
        //    $b1 = round((($fore_color & 0x00FF00) >> 8), 5);
         //   $g1 = round(($fore_color & 0x0000FF), 5);

            // convert a hexadecimal color code into decimal eps format (green = 0 1 0, blue = 0 0 1, ...)
           // $r2 = round((($back_color & 0xFF0000) >> 16), 5);
           // $b2 = round((($back_color & 0x00FF00) >> 8), 5);
           // $g2 = round(($back_color & 0x0000FF), 5);

            $color_mn = str_split($back_color, 2);
            $bincolor[0] = hexdec("0x{$color_mn[0]}");
            $bincolor[1] = hexdec("0x{$color_mn[1]}");
            $bincolor[2] = hexdec("0x{$color_mn[2]}");

            $color_bk = str_split($fore_color, 2);
            $bincolor2[0] = hexdec("0x{$color_bk[0]}");
            $bincolor2[1] = hexdec("0x{$color_bk[1]}");
            $bincolor2[2] = hexdec("0x{$color_bk[2]}");


            //$col[0] = ImageColorAllocate($base_image,$r2,$b2,$g2);
            //$col[1] = ImageColorAllocate($base_image,$r1,$b1,$g1);

            $col[0] = ImageColorAllocate($base_image, $bincolor2[0], $bincolor2[1], $bincolor2[2]);
            $col[1] = ImageColorAllocate($base_image,$bincolor[0], $bincolor[1], $bincolor[2]);

            imagefill($base_image, 0, 0, $col[0]);

            for($y=0; $y<$h; $y++) {
                for($x=0; $x<$w; $x++) {
                    if ($frame[$y][$x] == '1') {
                        ImageSetPixel($base_image,$x+$outerFrame,$y+$outerFrame,$col[1]);
                    }
                }
            }

            $target_image =ImageCreate($imgW * $pixelPerPoint, $imgH * $pixelPerPoint);
            ImageCopyResized($target_image, $base_image, 0, 0, 0, 0, $imgW * $pixelPerPoint, $imgH * $pixelPerPoint, $imgW, $imgH);
            ImageDestroy($base_image);

            return $target_image;
        }
    }


qr_code/pi.qr_code.php
<?php
namespace Mx\Qr_code;

require_once __DIR__.'/vendor/autoload.php';

use Endroid\QrCode\ErrorCorrectionLevel;
use Endroid\QrCode\QrCode;


/**
 * MX QR Code plugin for ExpressionEngine CMS 5.x
 *
 * MX QR Code helps you to generate qr codes.
 *
 * @link      https://eec.ms
 * @copyright Copyright (c) 2019 Max Lazar
 */


$plugin_info = array(
    'pi_name' => 'MX QR code',
    'pi_version' => '3.0.1',
    'pi_author' => 'Max Lazar',
    'pi_author_url' => 'http://eec.ms',
    'pi_description' => 'QR Code Generator',
    'pi_usage' => Qr_code::usage(),
);

class Qr_code
{
    public $return_data = '';

    private $ecc_levels = array(
            'L' => 'low',
            'M' => 'medium',
            'Q' => 'quartile',
            'H' => 'high',
        );

    private $_cache_path = '';

    /**
     * Package name.
     *
     * @var string
     */
    protected $package;

    /**
     * [__construct description].
     */
    public function __construct()
    {
        $this->_cache_path = (!$this->_cache_path) ? str_replace('\\', '/', PATH_CACHE).'/email_from_template' : false;
        $this->package = basename(__DIR__);
        $this->info = ee('App')->get($this->package);

        $data = array(
            'd' => (!ee()->TMPL->fetch_param('data')) ? ee()->TMPL->tagdata : str_replace(SLASH, '/', ee()->TMPL->fetch_param('data')),
            'e' => (!ee()->TMPL->fetch_param('ecc')) ? 'medium' :
                    (isset($this->ecc_levels[strtoupper(ee()->TMPL->fetch_param('ecc'))]) ? $this->ecc_levels[strtoupper(ee()->TMPL->fetch_param('ecc'))] :
                        'medium'),
            't' => (!ee()->TMPL->fetch_param('type')) ? 'PNG' : ee()->TMPL->fetch_param('type'),
            's' => (!ee()->TMPL->fetch_param('size')) ? '200' : ee()->TMPL->fetch_param('size'),
            'v' => (!ee()->TMPL->fetch_param('version')) ? null : ee()->TMPL->fetch_param('version'),
        );

        $action = (!ee()->TMPL->fetch_param('action')) ? ee()->TMPL->tagdata : ee()->TMPL->fetch_param('action');

        $data['bk_color'] = (ee()->TMPL->fetch_param('bk_color')) ? ltrim(ee()->TMPL->fetch_param('bk_color'), '#') : 'ffffff';
        $data['px_color'] = (ee()->TMPL->fetch_param('px_color')) ? ltrim(ee()->TMPL->fetch_param('px_color'), '#') : '000000';
        $data['bk_opacity'] = (ee()->TMPL->fetch_param('bk_opacity')) ? ee()->TMPL->fetch_param('bk_opacity') : '1';
        $data['px_opacity'] = (ee()->TMPL->fetch_param('px_opacity')) ? ee()->TMPL->fetch_param('px_opacity') : '1';

        $data['outline_size'] = (ee()->TMPL->fetch_param('outline_size')) ? ee()->TMPL->fetch_param('outline_size') : 2;
        $data['margin'] = (ee()->TMPL->fetch_param('margin')) ? ee()->TMPL->fetch_param('margin') : 0;
        $data['logo'] = (ee()->TMPL->fetch_param('logo')) ? ee()->TMPL->fetch_param('logo') : false;
        $data['logo_size'] = (ee()->TMPL->fetch_param('logo_size')) ? ee()->TMPL->fetch_param('logo_size') : false;
        $data['base64_encode'] = ee()->TMPL->fetch_param('base64_encode', 'no');

        switch ($action) {
        case 'sms':
            $tel = (!ee()->TMPL->fetch_param('tel')) ? '' : ee()->TMPL->fetch_param('tel');
            $data['d'] = 'SMSTO:'.((!ee()->TMPL->fetch_param('tel')) ? '' : ee()->TMPL->fetch_param('tel')).':'.$data['d'];
            break;
        case 'email':
            $data['d'] = 'SMTP:'.((!ee()->TMPL->fetch_param('email')) ? '' : ee()->TMPL->fetch_param('email')).':'.((!ee()->TMPL->fetch_param('sabj')) ? '' : ee()->TMPL->fetch_param('sabj')).':'.$data['d'];
            break;
        case 'tel':
            $data['d'] = 'TEL:'.((!ee()->TMPL->fetch_param('tel')) ? '' : ee()->TMPL->fetch_param('tel'));
            break;
        case 'site':
            $data['d'] = $this->SmartUrlEncode($data['d']);
            break;
        case 'bm':
            $data['d'] = 'MEBKM:TITLE:'.((!ee()->TMPL->fetch_param('title')) ? '' : ee()->TMPL->fetch_param('title')).':'.urlencode($data['d']);
            break;
        }

        return $this->return_data = $this->_doQRcode($data);
    }

    private function _doQRcode($data)
    {
        $output = '';
        $base_path = (!ee()->TMPL->fetch_param('base_path')) ? $_SERVER['DOCUMENT_ROOT'].'/' : ee()->TMPL->fetch_param('base_path');
        $base_path = str_replace('\\', '/', $base_path);
        $base_path = reduce_double_slashes($base_path);
        $cache = (!ee()->TMPL->fetch_param('cache')) ? '' : ee()->TMPL->fetch_param('cache');

        $base_cache = reduce_double_slashes($base_path.'images/cache/');
        $base_cache = (!ee()->TMPL->fetch_param('base_cache')) ? $base_cache : ee()->TMPL->fetch_param('base_cache');
        $base_cache = reduce_double_slashes($base_cache);

        if (!is_dir($base_cache)) {
            // make the directory if we can
            if (!mkdir($base_cache, 0777, true)) {
                ee()->TMPL->log_item('Error: could not create cache directory '.$base_cache.' with 777 permissions');

                return ee()->TMPL->no_results();
            }
        }

        $file_ext = ('J' == $data['t'] ? '.jpeg' : '.png');
        $file_name = md5(serialize($data)).$file_ext;

        if (!is_readable($base_cache.$file_name)) {
            $qrcode_data_string = $data['d'];
            $qrcode_error_correct = $data['e'];
            $qrcode_module_size = $data['s'];
            $qrcode_version = $data['v'];
            $qrcode_image_type = $data['t'];

            $qrCode = new QrCode($qrcode_data_string);

            $qrCode->setSize($data['s']);
            $qrCode->setMargin($data['margin']);
            $qrCode->setErrorCorrectionLevel(new ErrorCorrectionLevel($data['e']));
            $qrCode->setForegroundColor($this->hex2rgba($data['px_color'], $data['px_opacity']));
            $qrCode->setBackgroundColor($this->hex2rgba($data['bk_color'], $data['bk_opacity']));

            if ($data['logo']) {
                $qrCode->setLogoPath($this->url2local($data['logo']));
                if ($data['logo_size']) {
                    $size = explode(',', $data['logo_size']);
                    if (2 == count($size)) {
                        $qrCode->setLogoSize($size[0], $size[1]);
                    }
                }
            }

            if ('yes' == $data['base64_encode']) {
                $output = 'data:image/png;base64, '.base64_encode($qrCode->writeString());
            } else {
                $qrCode->writeFile($base_cache.$file_name);
                $output = reduce_double_slashes('/'.str_replace($base_path, '', $base_cache.$file_name));
            }
        } else {
            $output = reduce_double_slashes('/'.str_replace($base_path, '', $base_cache.$file_name));
        }


        return $output;
    }

    /**
     * [SmartUrlEncode description].
     *
     * @param [type] $url [description]
     */
    private function SmartUrlEncode($url)
    {
        if (false === strpos($url, '=')):
            return $url; else:
            $startpos = strpos($url, '?');
        $tmpurl = substr($url, 0, $startpos + 1);
        $qryStr = substr($url, $startpos + 1);
        $qryvalues = explode('&', $qryStr);
        foreach ($qryvalues as $value):
            $buffer = explode('=', $value);
        $buffer[1] = urlencode($buffer[1]);
        endforeach;
        $finalqrystr = implode('&amp;', $qryvalues);
        $finalURL = $tmpurl.$finalqrystr;

        return $finalURL;
        endif;
    }

    public function url2local($url)
    {
        $this->log_debug_message('Load remote file', 'start');
        //Download the file using file_get_contents.
        $downloadedFileContents = file_get_contents($url);

        //Check to see if file_get_contents failed.
        if (false === $downloadedFileContents) {
            throw new Exception('Failed to download file at: '.$url);
        }

        $extension = pathinfo(parse_url($url, PHP_URL_PATH), PATHINFO_EXTENSION);
        $file_name = $this->_cache_path.'/'.time().'.'.$extension;

        file_put_contents($file_name, $downloadedFileContents);

        return $file_name;
    }

    /* Convert hexdec color string to rgb(a) string */

    public function hex2rgba($color, $opacity = false)
    {
        $output = ['r' => 0, 'g' => 0, 'b' => 0, 'a' => 0];

        //Return default if no color provided
        if (empty($color)) {
            return $output;
        }

        //Sanitize $color if "#" is provided
        if ('#' == $color[0]) {
            $color = substr($color, 1);
        }

        //Check if color has 6 or 3 characters and get values
        if (6 == strlen($color)) {
            $hex = array($color[0].$color[1], $color[2].$color[3], $color[4].$color[5]);
        } elseif (3 == strlen($color)) {
            $hex = array($color[0].$color[0], $color[1].$color[1], $color[2].$color[2]);
        } else {
            return $output;
        }

        //Convert hexadec to rgb
        $rgb = array_map('hexdec', $hex);

        //Check if opacity is set(rgba or rgb)
        if ($opacity) {
            if (abs($opacity) > 1) {
                $opacity = 1.0;
            }
            $output['a'] = $opacity;
        }

        $output['r'] = $rgb[0];
        $output['g'] = $rgb[1];
        $output['b'] = $rgb[2];
        //Return rgb(a) color string
        return $output;
    }

    /**
     * Simple method to log a debug message to the EE Debug console.
     *
     * @param string $method
     * @param string $message
     */
    protected function log_debug_message($method = '', $message = '')
    {
        ee()->TMPL->log_item('&nbsp;&nbsp;***&nbsp;&nbsp;'.$this->package." - $method debug: ".$message);
    }

    // ----------------------------------------
    //  Plugin Usage
    // ----------------------------------------

    // This function describes how the plugin is used.
    //  Make sure and use output buffering

    public static function usage()
    {
        // for performance only load README if inside control panel
        return REQ === 'CP' ? file_get_contents(dirname(__FILE__).'/README.md') : null;
    }

    /* END */
}
